context.go

```diff
@@ -75,52 +77,69 @@ func (d RunAddressData) isEmpty() bool {
 // matches and actions can still be returned, for instance in the case of a timeout error. Errors can be tested against
 // the RunError type.
 // Struct fields having the tag `ddwaf:"ignore"` will not be encoded and sent to the WAF
-func (context *Context) Run(addressData RunAddressData, timeout time.Duration) (res Result, err error) {
+// if the output of TotalTime() exceeds the value of TimeBudget, the function will immediately return with an error
+func (context *Context) Run(addressData RunAddressData, _ time.Duration) (res Result, err error) {
 	if addressData.isEmpty() {
-		return
+		return res, nil
 	}
 
-	now := time.Now()
+	// Everything here is thread safe and ran after context.mutex.Unlock() so we lock the context the less possible
+	// so the code here needs to be thread safe
 	defer func() {
-		dt := time.Since(now)
-		context.totalOverallRuntimeNs.Add(uint64(dt.Nanoseconds()))
+		if stdErrors.Is(err, errors.ErrTimeout) {
+			context.timeBudgetExpirationCount.Inc()
+		}
 	}()
 
-	// At this point, the only error we can get is an error in case the top level object is a nil map, but this
-	// behaviour is expected since either persistent or ephemeral addresses are allowed to be null one at a time.
-	// In this case, EncodeAddresses will return nil contrary to Encode which will return an nil wafObject,
-	// which is what we need to send to ddwaf_run to signal that the address data is empty.
-	var persistentData *bindings.WafObject = nil
-	var ephemeralData *bindings.WafObject = nil
-	persistentEncoder := newLimitedEncoder()
-	ephemeralEncoder := newLimitedEncoder()
-	if addressData.Persistent != nil {
-		persistentData, _ = persistentEncoder.EncodeAddresses(addressData.Persistent)
+	// If the context has already timed out, we don't need to run the WAF again
+	if context.timings.Expired() {
+		return res, errors.ErrTimeout
 	}
 
-	if addressData.Ephemeral != nil {
-		ephemeralData, _ = ephemeralEncoder.EncodeAddresses(addressData.Ephemeral)
+	runTimings := context.timings.newRun()
+	startTime := time.Now()
+	defer func() {
+		runTimings.RunNs.Add(time.Since(startTime).Nanoseconds())
+		context.timings.absorb(&runTimings)
```

Comment:
I think this misbehaves if `context.Run()` gets called more than once per context.
`timings.newRun()` creates a `Timings` where the run duration is the current value of context.timings.
This works fine when no run has occured, but if one has already occurred and the value is not zero then `absorb` will end up adding too much to the context timings run duration (it will basically do context run duration * 2 + run duration) .